File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    boolean isSameTree = true;
    boolean result = false;
    
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        // approach: DFS -> when roots are same, call same tree
        
        dfs(root, subRoot);
        return result;
    }
    
    private void dfs(TreeNode root, TreeNode subRoot) {
        if (root == null || subRoot == null) {
            return;
        }
        
        if (root.val == subRoot.val) {
            isSameTree = true;
            helper(root, subRoot);
            
            if (isSameTree) {
                result = true;
            }
        }
        
        dfs(root.left, subRoot);
        dfs(root.right, subRoot);
    }
    
    private void helper(TreeNode p, TreeNode q) {
        if (p == null && q == null) {
            return;
        }
        
        if (p == null || q == null) {
            isSameTree = false;
            return;
        }
        
        if (p.val != q.val) {
            isSameTree = false;
        }
        
        helper(p.left, q.left);
        helper(p.right, q.right);
    }
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        # approach: top-down DFS, if roots match, call same tree
        
        isSameTree = True
        result = False
        
        def helper(p, q):
            nonlocal isSameTree
            nonlocal result
            
            if not p or not q:
                return
            
            if p.val == q.val:
                isSameTree = True
                sameTree(p, q)
                
                if isSameTree:
                    result = True
            
            helper(p.left, q)
            helper(p.right, q)
        
        def sameTree(p, q):
            nonlocal isSameTree
            
            if not p and not q:
                return
            
            if not p or not q:
                isSameTree = False
                return

            if p.val != q.val:
                isSameTree = False
            
            sameTree(p.left, q.left)
            sameTree(p.right, q.right)
        
        helper(root, subRoot)
        return result
        